File: IsSubsequence2.py
class Solution:
    def isSubsequence(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        def binarySearch(nums, target):
            l, r = 0, len(nums) - 1
            while l <= r:
                mid = l + (r - l) // 2
                if nums[mid] >= target:
                    r = mid - 1
                else:
                    l = mid + 1
            return l

        m = {}
        for i, ch in enumerate(t):
            if ch not in m:
                m[ch] = []
            m[ch].append(i)
        # for i in range(len(t)):
        #     if t[i] not in m:
        #         m[t[i]] = []
        #     m[t[i]].append(i)

        pre = 0
        for ch in s:
            if ch not in m:
                return False
            index = binarySearch(m[ch], pre)
            if index >= len(m[ch]):
                return False
            pre = m[ch][index] + 1
        return True


if __name__ == '__main__':
    s = input()
    t = input()
    solution = Solution()
    print(solution.isSubsequence(s, t))

File: IsSubsequence2.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    bool isSubsequence(string s, string t) {
        
        vector<vector<int> > m(26);
        
        for(int i = 0; i < t.size(); i++) {
        	m[t[i] - 'a'].push_back(i);
		}
		
		int pre = 0;
		for(char c : s) {
			if(m[c - 'a'].empty()) return false;
			int index = binarySearch(m[c - 'a'], pre);
			cout<<index<<endl;
			if(index >= m[c - 'a'].size()) return false;
			pre = m[c - 'a'][index] + 1;
		}
        
        return true;
    }
    
    int binarySearch(vector<int> nums, int target) {
    	int l = 0, r = nums.size() - 1;
    	while(l <= r) {
    		int mid = l + (r - l) / 2;
    		if(nums[mid] >= target) r = mid - 1;
    		else l = mid + 1;
		}
		return l;
	}
};

int main() {
	
	string s, t;
	cin>>s;
	cin>>t;
	
	Solution *solution = new Solution();
	solution->isSubsequence(s, t) ? cout<<"true" : cout<<"false";
	
	return 0;
	
}
